### DISCUSSION TITLE
使用umiMax数据流管理插件时，getInitialState和model中useEffect代码执行顺序相关问题

### DISCUSSION THREAD

**Discussion by Wei0809** [umijs/umi#11870] [Q&A]
例如[该示例](https://codesandbox.io/p/github/Wei0809/umi-Example/master?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cloqx4tky00vq3b6hjmdqvrpt%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cloqx4tky00vn3b6hwsbp7nf7%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cloqx4tky00vp3b6hp16p65h9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cloqx4tky00vo3b6hpwez4lfp%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cloqx4tky00vn3b6hwsbp7nf7%2522%253A%257B%2522id%2522%253A%2522cloqx4tky00vn3b6hwsbp7nf7%2522%252C%2522activeTabId%2522%253A%2522clozj4s16006x3b6h7qhy0agv%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fapp.ts%2522%252C%2522id%2522%253A%2522clozj4s16006x3b6h7qhy0agv%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cloqx4tky00vo3b6hpwez4lfp%2522%253A%257B%2522id%2522%253A%2522cloqx4tky00vo3b6hpwez4lfp%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cloqx4tky00vp3b6hp16p65h9%2522%253A%257B%2522id%2522%253A%2522cloqx4tky00vp3b6hp16p65h9%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cloqwz21x00043b6hj3qzafn9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cloqwz2dw001jeei26dhn641i%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522cloqwz34i007d3b6h42v5z2t5%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522cloqx0ypl00bf3b6hchheo7dv%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cloqwz34i007d3b6h42v5z2t5%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)中，我加了打印后发现global model的useEffect中的代码是最先执行的，然后才是getInitialState方法，请问有什么方法可以让getInitialState执行完毕后，别的model才执行吗？
![image](https://github.com/umijs/umi/assets/43309547/c5be8b07-6b72-48be-9ec0-a1d7e2454bb5)
![image](https://github.com/umijs/umi/assets/43309547/5ad3aa4f-694a-4b2b-a4b6-c1b2105a4b58)
![image](https://github.com/umijs/umi/assets/43309547/d33345b0-19c1-4b6d-a472-894f14b91c2a)

**Comment by fz6m** [comment]
umi 里的 `useModel` 这个功能是全局的一个 react fiber 节点来提供的持久化，里面的副作用都是一开始就会执行的，所以不建议把副作用写到 model 里，如果你想让某个副作用在某个区域内才执行，应该把他放到对应组件区域的 `useEffect` 或者 context 里去执行；社区里有一个 [hox](https://github.com/umijs/hox) 库，可以创建一个局部持久的 model ，如有需要可以自行取用。

`getInitialState` 也是一个 model ，如同上面所说，model 都是全局持久的，所以不要包含副作用，或者你就在你的 model 里 `useModel('@@initialState')` 取一下 `getInitialState` 的返回数据，`useEffect` 监听一下，等 `getInitialState` 获取到了再去执行后续的逻辑。
